File: react-native/src/tracing/msrSpan.ts
import type { ValidAttributeValue } from "../utils/attributeValueValidator";
import type { IIdProvider } from "../utils/idProvider";
import type { TimeProvider } from "../utils/timeProvider";
import type { Checkpoint } from "./checkpoint";
import type { InternalSpan } from "./internalSpan";
import type { Span } from "./span";
import type { SpanData } from "./spanData";
import type { ISpanProcessor } from "./spanProcessor";
import { EndState, SpanStatus } from "./spanStatus";
import type { ITraceSampler } from "./traceSampler";

export class MsrSpan implements InternalSpan {
  timeProvider: TimeProvider;
  spanProcessor: ISpanProcessor;

  isSampled: boolean;
  name: string;
  spanId: string;
  traceId: string;
  parentId: string | undefined;
  startTime: number;

  status: SpanStatus = SpanStatus.Unset;
  endTime: number = 0;
  hasEndedState: EndState = EndState.NotEnded;
  checkpoints: Checkpoint[] = [];
  attributes: { [key: string]: any } | undefined;
  userDefinedAttrs: Record<string, ValidAttributeValue> = {};

  constructor(
    timeProvider: TimeProvider,
    isSampled: boolean,
    name: string,
    spanId: string,
    traceId: string,
    parentId: string | undefined,
    startTime: number,
    spanProcessor: ISpanProcessor
  ) {
    this.timeProvider = timeProvider;
    this.isSampled = isSampled;
    this.name = name;
    this.spanId = spanId;
    this.traceId = traceId;
    this.parentId = parentId;
    this.startTime = startTime;
    this.spanProcessor = spanProcessor;
  }

  static startSpan({
    name,
    timeProvider,
    idProvider,
    traceSampler,
    parentSpan,
    spanProcessor,
    timestamp = timeProvider.now(),
  }: {
    name: string;
    timeProvider: TimeProvider;
    idProvider: IIdProvider;
    traceSampler: ITraceSampler;
    parentSpan?: Span;
    spanProcessor: ISpanProcessor;
    timestamp?: number;
  }): Span {
    const startTime = timestamp;
    const spanId = idProvider.spanId();
    const traceId = parentSpan?.traceId ?? idProvider.traceId();
    const isSampled = parentSpan?.isSampled ?? traceSampler.shouldSampleTrace(traceId);

    const span = new MsrSpan(
      timeProvider,
      isSampled,
      name,
      spanId,
      traceId,
      parentSpan?.spanId,
      startTime,
      spanProcessor
    );

    spanProcessor.onStart(span);
    return span;
  }

  getStatus(): SpanStatus {
    return this.status;
  }

  getUserDefinedAttrs(): Record<string, ValidAttributeValue> {
    return this.userDefinedAttrs;
  }

  setInternalAttribute(attribute: { [key: string]: any }): void {
    this.attributes = attribute;
  }

  setStatus(status: SpanStatus): Span {
    if (this.hasEndedState === EndState.NotEnded) {
      this.status = status;
    }
    return this;
  }

  setParent(parentSpan: Span): Span {
    if (this.hasEndedState === EndState.NotEnded) {
      this.parentId = parentSpan.spanId;
      this.traceId = parentSpan.traceId;
    }
    return this;
  }

  setCheckpoint(name: string): Span {
    if (this.hasEndedState === EndState.NotEnded) {
      const timestamp = this.timeProvider.now();
      const checkpoint: Checkpoint = { name, timestamp };
      this.checkpoints.push(checkpoint);
    }
    return this;
  }

  setName(name: string): Span {
    if (this.hasEndedState === EndState.NotEnded) {
      this.name = name;
    }
    return this;
  }

  setAttribute(key: string, value: ValidAttributeValue): Span {
    if (this.hasEndedState === EndState.NotEnded) {
      this.userDefinedAttrs[key] = value;
    }
    return this;
  }

  setAttributes(attributes: Record<string, ValidAttributeValue>): Span {
    if (this.hasEndedState === EndState.NotEnded) {
      Object.assign(this.userDefinedAttrs, attributes);
    }
    return this;
  }

  removeAttribute(key: string): Span {
    if (this.hasEndedState === EndState.NotEnded) {
      delete this.userDefinedAttrs[key];
    }
    return this;
  }

  end(timestamp?: number): Span {
    if (this.hasEndedState !== EndState.NotEnded) {
      return this;
    }

    this.hasEndedState = EndState.Ending;
    this.endTime = timestamp ?? this.timeProvider.now();
    this.spanProcessor.onEnding(this);

    this.hasEndedState = EndState.Ended;
    this.spanProcessor.onEnded(this);
    return this;
  }

  hasEnded(): boolean {
    return this.hasEndedState === EndState.Ended;
  }

  getDuration(): number {
    return this.calculateDuration();
  }

  setSampled(sampled: boolean): void {
    this.isSampled = sampled;
  }

  toSpanData(): SpanData {
    return {
      name: this.name,
      traceId: this.traceId,
      spanId: this.spanId,
      parentId: this.parentId,
      startTime: this.startTime,
      endTime: this.endTime,
      duration: this.calculateDuration(),
      status: this.status,
      attributes: this.attributes,
      userDefinedAttrs: this.userDefinedAttrs,
      checkpoints: this.checkpoints,
      hasEnded: this.hasEndedState === EndState.Ended,
      isSampled: this.isSampled,
    };
  }

  private calculateDuration(): number {
    if (this.hasEndedState === EndState.Ended) {
      return this.endTime - this.startTime;
    }
    return 0;
  }
}

File: android/measure-android/measure/src/main/java/sh/measure/android/tracing/MsrSpan.kt
package sh.measure.android.tracing

import sh.measure.android.SessionManager
import sh.measure.android.attributes.AttributeValue
import sh.measure.android.logger.LogLevel
import sh.measure.android.logger.Logger
import sh.measure.android.utils.IdProvider
import sh.measure.android.utils.Sampler
import sh.measure.android.utils.TimeProvider

/**
 * A thread safe implementation of [Span].
 */
internal class MsrSpan(
    private val logger: Logger,
    private val timeProvider: TimeProvider,
    private val spanProcessor: SpanProcessor,
    override var isSampled: Boolean,
    override var name: String,
    override val spanId: String,
    override var traceId: String,
    override var parentId: String?,
    override val sessionId: String,
    override val startTime: Long,
) : InternalSpan {
    private val lock = Any()
    private var status = SpanStatus.Unset
    private var endTime = 0L
    private var hasEnded: EndState = EndState.NotEnded
    override val checkpoints: MutableList<Checkpoint> = mutableListOf()
    override val attributes: MutableMap<String, Any?> = mutableMapOf()
    private val userDefinedAttrs: MutableMap<String, Any?> = mutableMapOf()

    companion object {
        fun startSpan(
            name: String,
            logger: Logger,
            timeProvider: TimeProvider,
            spanProcessor: SpanProcessor,
            sessionManager: SessionManager,
            idProvider: IdProvider,
            sampler: Sampler,
            parentSpan: Span?,
            timestamp: Long? = null,
        ): Span = InternalTrace.trace(
            { "msr-startSpan" },
            {
                val startTime = timestamp ?: timeProvider.now()
                val spanId: String = idProvider.spanId()
                val traceId = parentSpan?.traceId ?: idProvider.traceId()
                val sessionId = sessionManager.getSessionId()
                val isSampled = parentSpan?.isSampled ?: sampler.shouldSampleTrace(traceId)
                val span = MsrSpan(
                    logger = logger,
                    timeProvider = timeProvider,
                    spanProcessor = spanProcessor,
                    name = name,
                    spanId = spanId,
                    traceId = traceId,
                    parentId = parentSpan?.spanId,
                    sessionId = sessionId,
                    startTime = startTime,
                    isSampled = isSampled,
                )
                spanProcessor.onStart(span)
                span
            },
        )
    }

    override fun getStatus(): SpanStatus {
        synchronized(lock) {
            return this.status
        }
    }

    override fun setStatus(status: SpanStatus): Span {
        synchronized(lock) {
            if (hasEnded != EndState.NotEnded) {
                logger.log(
                    LogLevel.Error,
                    "Failed to update span: attempt to set status after span ended",
                )
                return this
            }
            this.status = status
        }
        return this
    }

    override fun setParent(parentSpan: Span): Span {
        synchronized(lock) {
            if (hasEnded != EndState.NotEnded) {
                logger.log(
                    LogLevel.Error,
                    "Failed to update span: attempt to set parent after span ended",
                )
                return this
            }
            this.parentId = parentSpan.spanId
            this.traceId = parentSpan.traceId
        }
        return this
    }

    override fun setCheckpoint(name: String): Span {
        synchronized(lock) {
            if (hasEnded != EndState.NotEnded) {
                logger.log(
                    LogLevel.Error,
                    "Failed to update span: attempt to set checkpoint after span ended",
                )
                return this
            }
            val checkpoint = Checkpoint(name, timeProvider.now())
            this.checkpoints.add(checkpoint)
        }
        return this
    }

    override fun setName(name: String): Span {
        synchronized(lock) {
            if (hasEnded != EndState.NotEnded) {
                logger.log(
                    LogLevel.Error,
                    "Failed to update span: attempt to set name after span ended",
                )
                return this
            }
            this.name = name
        }
        return this
    }

    override fun setSamplingRate(sampled: Boolean) {
        synchronized(lock) {
            // Allow updating ended spans too
            // to account for lazy sampling
            // config loading.
            this.isSampled = sampled
        }
    }

    override fun setAttribute(key: String, value: String): Span {
        synchronized(lock) {
            if (hasEnded != EndState.NotEnded) {
                logger.log(
                    LogLevel.Error,
                    "Failed to update span: attempt to set attribute after span ended",
                )
                return this
            }
            this.userDefinedAttrs[key] = value
        }
        return this
    }

    override fun setAttribute(key: String, value: Long): Span {
        synchronized(lock) {
            if (hasEnded != EndState.NotEnded) {
                logger.log(
                    LogLevel.Error,
                    "Failed to update span: attempt to set attribute after span ended",
                )
                return this
            }
            this.userDefinedAttrs[key] = value
        }
        return this
    }

    override fun setAttribute(key: String, value: Int): Span {
        synchronized(lock) {
            if (hasEnded != EndState.NotEnded) {
                logger.log(
                    LogLevel.Error,
                    "Failed to update span: attempt to set attribute after span ended",
                )
                return this
            }
            this.userDefinedAttrs[key] = value
        }
        return this
    }

    override fun setAttribute(key: String, value: Double): Span {
        synchronized(lock) {
            if (hasEnded != EndState.NotEnded) {
                logger.log(
                    LogLevel.Error,
                    "Failed to update span: attempt to set attribute after span ended",
                )
                return this
            }
            this.userDefinedAttrs[key] = value
        }
        return this
    }

    override fun setAttribute(key: String, value: Boolean): Span {
        synchronized(lock) {
            if (hasEnded != EndState.NotEnded) {
                logger.log(
                    LogLevel.Error,
                    "Failed to update span: attempt to set attribute after span ended",
                )
                return this
            }
            this.userDefinedAttrs[key] = value
        }
        return this
    }

    override fun setAttributes(attributes: Map<String, AttributeValue>): Span {
        synchronized(lock) {
            if (hasEnded != EndState.NotEnded) {
                logger.log(
                    LogLevel.Error,
                    "Failed to update span: attempt to set attribute after span ended",
                )
                return this
            }
            attributes.forEach { (key, value) ->
                this.userDefinedAttrs[key] = value.value
            }
        }
        return this
    }

    override fun removeAttribute(key: String): Span {
        synchronized(lock) {
            if (hasEnded != EndState.NotEnded) {
                logger.log(
                    LogLevel.Error,
                    "Failed to update span: attempt to set attribute after span ended",
                )
                return this
            }
            this.userDefinedAttrs.remove(key)
        }
        return this
    }

    override fun end(): Span {
        endSpanInternal(timeProvider.now())
        return this
    }

    override fun end(timestamp: Long): Span {
        endSpanInternal(timestamp)
        return this
    }

    override fun hasEnded(): Boolean {
        synchronized(lock) {
            return hasEnded != EndState.NotEnded
        }
    }

    private fun endSpanInternal(timestamp: Long) {
        synchronized(lock) {
            if (hasEnded != EndState.NotEnded) {
                logger.log(
                    LogLevel.Error,
                    "Failed to update span: attempt to end and already ended span",
                )
                return
            }
            endTime = timestamp
            hasEnded = EndState.Ending
        }
        spanProcessor.onEnding(this)
        synchronized(lock) {
            hasEnded = EndState.Ended
        }
        spanProcessor.onEnded(this)
    }

    override fun getDuration(): Long {
        synchronized(lock) {
            if (hasEnded != EndState.Ended) {
                logger.log(
                    LogLevel.Error,
                    "Failed to get duration of a span($name): span has not ended",
                )
                return 0
            } else {
                return calculateDuration()
            }
        }
    }

    override fun setInternalAttribute(attribute: Pair<String, Any?>) {
        synchronized(lock) {
            if (hasEnded != EndState.Ended) {
                attributes[attribute.first] = attribute.second
            } else {
                logger.log(
                    LogLevel.Error,
                    "Failed to update span: attempt to set attribute after span ended",
                )
            }
        }
    }

    override fun getAttributesMap(): MutableMap<String, Any?> {
        synchronized(lock) {
            return attributes
        }
    }

    override fun getUserDefinedAttrs(): MutableMap<String, Any?> {
        synchronized(lock) {
            return userDefinedAttrs
        }
    }

    override fun toSpanData(): SpanData {
        synchronized(lock) {
            return SpanData(
                spanId = spanId,
                traceId = traceId,
                name = name,
                startTime = startTime,
                endTime = endTime,
                status = status,
                hasEnded = hasEnded == EndState.Ended,
                parentId = parentId,
                sessionId = sessionId,
                checkpoints = checkpoints,
                attributes = attributes,
                userDefinedAttrs = userDefinedAttrs,
                duration = calculateDuration(),
                isSampled = isSampled,
            )
        }
    }

    private fun calculateDuration(): Long = endTime - startTime

    private enum class EndState {
        NotEnded,
        Ending,
        Ended,
    }
}
